File: Stack/GFG/Medium/PostfixtoInfix/PostfixtoInfix.cpp
// https://www.geeksforgeeks.org/problems/postfix-to-infix-conversion/1
#include<bits/stdc++.h>
using namespace std;
string postToInfix(string &exp)
{
    // Write your code here
    stack<string> st;
    for (char &ch : exp)
    {
        if (isalnum(ch))
        {
            st.push(string(1, ch));
        }
        else
        {
            string right = st.top();
            st.pop();
            string left = st.top();
            st.pop();
            st.push("(" + left + ch + right + ")");
        }
    }
    return st.top();
}

File: Stack/GFG/Medium/PostfixtoInfix/PostfixtoInfix.java
// https://www.geeksforgeeks.org/problems/postfix-to-infix-conversion/1

import java.util.ArrayDeque;
import java.util.Deque;

public class PostfixtoInfix {
    static String postToInfix(String exp) {
        // code here
        String str = "";
        Deque<String> st = new ArrayDeque<>();
        
        for(int i=0; i<exp.length(); i++) {
            char ch = exp.charAt(i);
            if(Character.isLetterOrDigit(ch)) {
                st.push(""+ch);
            } else {
                String first = st.pop();
                String second = st.pop();
                String brac = "("+second+ch+first+")";
                st.push(brac);
            }
        }
        return st.pop();
    }
}
